File: src/tests/Behavioral/SelectRecvCommShadowRename/SelectRecvCommShadowRename.go
// A select whose comm clause is a RECEIVE ASSIGNMENT (`case x := <-c:`) must render the
// channel operand with the SAME shadow-renamed identifier every other reference uses.
//
// The variable-analysis pass visits a comm clause's expression only on the SEND and
// bare-RECEIVE arms; the ASSIGNMENT arm processed the LHS declarations alone, so no ident
// inside the RHS (`<-c`) ever received an identNames mapping. The select temp then emitted
// the RAW name — `var selᴛ1 = c;` — while the twelve other references in the region used the
// renamed `cΔ1`. That is CS0841 rather than CS0103 precisely because the rename was
// NECESSARY: a distinct, later `c` is declared in the same C# method scope, so the bare `c`
// binds forward to it.
//
// Guards reflect's TestChan (`all_test.go:1720`), whose `case x := <-c:` sits inside a loop
// body declaring `var c chan int` while a different `c` lives later in the same function.
package main

import "fmt"

func main() {
	for loop := 0; loop < 2; loop++ {
		// Shadow-renamed: the function declares a DIFFERENT `c` later on.
		var c chan int
		c = make(chan int, 2)
		c <- 40 + loop

		// Single-value receive assignment — the failing shape.
		select {
		case x := <-c:
			fmt.Println("recv", x)
		default:
			fmt.Println("empty")
		}

		// Two-value receive assignment — the same arm, comma-ok form.
		c <- 50 + loop
		select {
		case x, ok := <-c:
			fmt.Println("recv2", x, ok)
		default:
			fmt.Println("empty2")
		}

		// Nothing buffered: the default arm runs, proving the temp still binds.
		select {
		case x := <-c:
			fmt.Println("unexpected", x)
		default:
			fmt.Println("drained")
		}

		// A bare receive and a send in the same select keep working (the arms that
		// were already visited), mixed with the assignment arm above.
		d := make(chan int, 1)
		select {
		case d <- 7:
			fmt.Println("sent")
		case <-c:
			fmt.Println("unexpected recv")
		}
		fmt.Println("d", <-d)
	}

	// The LATER, DIFFERENT `c` whose presence forces the rename above.
	var c chan string
	c = make(chan string, 1)
	c <- "tail"
	fmt.Println(<-c)
}

File: src/tests/Behavioral/SelectRecvCommShadowRename/SelectRecvCommShadowRename.cs
namespace go;

using fmt = fmt_package;

partial class main_package {

// Go runs an imported package's `init` before this package's own; .NET would never load
// an assembly nothing has touched yet, so that initialization is forced here.
[GoInit] internal static void initᴛᴛimportꓸfmt() {
    builtin.initPackage(typeof(fmt_package));
}

// Hoisted @string literals (single allocation; Go keeps these in RODATA)
private static readonly object recvˢ = (@string)"recv"u8;
private static readonly object emptyˢ = (@string)"empty"u8;
private static readonly object recv2ˢ = (@string)"recv2"u8;
private static readonly object empty2ˢ = (@string)"empty2"u8;
private static readonly object unexpectedˢ = (@string)"unexpected"u8;
private static readonly object drainedˢ = (@string)"drained"u8;
private static readonly object sentˢ = (@string)"sent"u8;
private static readonly object unexpectedRecvˢ = (@string)"unexpected recv"u8;
private static readonly @string tailˢ = "tail"u8;

internal static void Main() {
    for (nint loop = 0; loop < 2; loop++) {
        channel<nint> cΔ1 = default!;
        cΔ1 = new channel<nint>(2);
        cΔ1.ᐸꟷ(40 + loop);
        var selᴛ1 = cΔ1;
        switch (trySelect(ᐸꟷ(selᴛ1, ꓸꓸꓸ))) {
        case 0 when selᴛ1.ꟷᐳ(out var x): {
            fmt.Println(recvˢ, x);
            break;
        }
        default: {
            fmt.Println(emptyˢ);
            break;
        }}
        cΔ1.ᐸꟷ(50 + loop);
        var selᴛ2 = cΔ1;
        switch (trySelect(ᐸꟷ(selᴛ2, ꓸꓸꓸ))) {
        case 0 when selᴛ2.ꟷᐳ(out var x, out var ok): {
            fmt.Println(recv2ˢ, x, ok);
            break;
        }
        default: {
            fmt.Println(empty2ˢ);
            break;
        }}
        var selᴛ3 = cΔ1;
        switch (trySelect(ᐸꟷ(selᴛ3, ꓸꓸꓸ))) {
        case 0 when selᴛ3.ꟷᐳ(out var x): {
            fmt.Println(unexpectedˢ, x);
            break;
        }
        default: {
            fmt.Println(drainedˢ);
            break;
        }}
        var d = new channel<nint>(1);
        var selᴛ4 = d.ᐸꟷ(7, ꓸꓸꓸ);
        var selᴛ5 = cΔ1;
        switch (select(selᴛ4, ᐸꟷ(selᴛ5, ꓸꓸꓸ))) {
        case 0: {
            fmt.Println(sentˢ);
            break;
        }
        case 1 when selᴛ5.ꟷᐳ(out _): {
            fmt.Println(unexpectedRecvˢ);
            break;
        }}
        fmt.Println((@string)"d"u8, ᐸꟷ(d));
    }
    channel<@string> c = default!;
    c = new channel<@string>(1);
    c.ᐸꟷ(tailˢ);
    fmt.Println(ᐸꟷ(c));
}

} // end main_package
